oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedMongoDownloadTask.java

```diff
@@ -101,12 +100,22 @@ public long getDocumentsDownloaded() {
      * filtering in the indexing job. This feature may significantly reduce the number of documents downloaded from
      * Mongo.
      * The performance gains may not be proportional to the reduction in the number of documents downloaded because Mongo
-     * still has to traverse all the documents. This is the case because the regex expression used for path filtering
+     * still has to traverse all the documents. This is required because the regex expression used for path filtering
      * starts with a wildcard (because the _id starts with the depth of the path, so the regex expression must ignore
      * this part). Because of the wildcard at the start, Mongo cannot use of the index on _id.
      */
     public static final String OAK_INDEXER_PIPELINED_MONGO_REGEX_PATH_FILTERING = "oak.indexer.pipelined.mongoRegexPathFiltering";
     public static final boolean DEFAULT_OAK_INDEXER_PIPELINED_MONGO_REGEX_PATH_FILTERING = false;
+
+    /**
+     * Maximum number of elements in the included/excluded paths list used for regex path filtering. If after
+     * merging and de-deduplication of the paths of all the path filters the number of included or excluded paths exceeds
+     * this value, then disable path filtering to avoid creating Mongo queries with large number of filters
+     */
+    public static final String OAK_INDEXER_PIPELINED_MONGO_REGEX_PATH_FILTERING_MAX_PATHS = "oak.indexer.pipelined.mongoRegexPathFilteringMaxPaths";
+    public static final int DEFAULT_OAK_INDEXER_PIPELINED_MONGO_REGEX_PATH_FILTERING_MAX_PATHS = 20;
```

Comment:
It would be good to verify what is the maximum number of entries that MongoDB supports... I could imagine it's less than 20 includedPath + 20 excludedPaths...

And I wonder if there is a limit on the length of the paths, not just the number of paths.

Reply:
The maximum size of a query is limited by the maximum size of a bson document, which is 16MB. I did not make a test myself, but I found several posts on StackOverflow that say that it is possible to have hundreds of fields in the arrays passed to `$in` and `$nin`. But we should anyway have a much lower limit in the tens because otherwise the performance would degrade.

I'm wondering if we should have hard limits to the number of included/excluded paths. Or just log a warning if there are more than, let's say, 100. I would go with a warning, because there may be valid reasons to have many included/excluded paths and Mongo might actually handle it correctly, so we do not want to restrict our freedom in the future. But a warning would already be useful, as if something would fail, the warning would point to the potential problem. What do you think?

Reply:
> The maximum size of a query is limited by the maximum size of a bson document, which is 16MB.

Great! 

A warning might make sense, but I don't think it's important to have a warning here. I think it's fine.